### PR TITLE
add transparency carving to nerfacto

### PR DESCRIPTION
**Pull request by SamDM**
there is a very simple trick not implemented in Nerfacto that significantly reduces floaters in synthetic scenes with few cameras. I implemented it, below is a small comparison with and without this addition.

Without:
plant-no-alpha-carve

With:
plant-with-alpha-carve

The trick only works if you have foreground masks, it works as follows:

1. Make everything except for the foreground object transparent in each image.
2. While rendering, pick a random color in each iteration, then use this random color as background during alpha compositing.
3. When computing the RGB loss, make the color of transparent pixels match the randomly chosen color of the current iteration used for rendering.

The end result is that floaters will get punished because then can't keep up with the constantly changing background color, therefore pushing them towards transparency.

### CODE REVIEW

**Comment by machenmusik** [comment]
Interesting.   How does this do with mustard dataset from https://github.com/nerfstudio-project/nerfstudio/issues/1498#issue-1599115766 I wonder...

**Comment by f-dy** [comment]
I think we should have the two kinds of masks:
1. ignore masks to ignore parts of the image (eg transients): do not sample rays outside of these masks
2. transparency masks to say everything outside of these masks should be transparent: adds the transparency loss

**Comment by nepfaff** [comment]
> I think we should have the two kinds of masks:
> 
> 1. ignore masks to ignore parts of the image (eg transients): do not sample rays outside of these masks
> 2. transparency masks to say everything outside of these masks should be transparent: adds the transparency loss

The first one is already implemented in Nerfstudio. The second one is what this PR is trying to do and what instantNGP is doing. I actually planned to implement something like this once I'm back from my vacation, so I'm excited that this PR already got opened. Thanks! I will try this on the masked mustard dataset soon and compare it with instantNGP.

**Comment by nepfaff** [comment]
@SamDM would you mind providing the data and training command that you used for your testing? It would be interesting to compare your results with my nerfacto results with default settings above.

**Comment by nepfaff** [comment]
I added the same logic as done in this PR to `NGPModel` and run it with the default settings of `instant-ngp-bounded`. The results now closely correspond to the ones when using `instantNGP` with alpha transparent training:

https://github.com/nerfstudio-project/nerfstudio/assets/53228351/b4fa76e2-216e-4df0-9bf7-429168ba58bb

**These results are great, and I would support merging this.**
 I'm still unsure why `nerfacto` did not work for this dataset when using alpha transparent training. If someone has any insights here, it would be great to mention something in the docs about when alpha transparent training is expected to work (model, parameters, etc.).

File: nerfstudio/models/nerfacto.py [hunk 1]
```diff
@@ -327,15 +322,20 @@ def get_outputs(self, ray_bundle: RayBundle):
...
     def get_loss_dict(self, outputs, batch, metrics_dict=None):
         loss_dict = {}
         image = batch["image"].to(self.device)
-        loss_dict["rgb_loss"] = self.rgb_loss(image, outputs["rgb"])
+        if outputs.get("background_color", None) is not None and image.shape[-1] == 4:
+            background_color = outputs["background_color"]
+            image_ = image[:, :3] * image[:, 3:] + background_color.to(image.device) * (1 - image[:, 3:])
+        else:
+            image_ = image[:, :3]
+        loss_dict["rgb_loss"] = self.rgb_loss(image_, outputs["rgb"])
```
Comment (nepfaff) [inline nerfstudio/models/nerfacto.py:338]: Can you add the same logic to [NGPModel](https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/models/instant_ngp.py)? I found that this model works (at least with the `instant-ngp-bounded` default settings) works particularly well with alpha transparent training (at least for the masked mustard dataset). It might also be beneficial for other models but maybe not worth adding for the others until it becomes useful for people.

**Comment by f-dy** [comment]
Just one concern: the option names and documentation. `use_alpha_channel` is a dataloader option, but we don't know what alpha is being used for. And the transparency loss is implemented only in nerfacto, and undocumented.

Alternative:
- alpha channel should be always loaded (so no use_alpha_channel option)
- models that don't implement transparency loss should discard it
- nerfacto should have an option to specifically use alpha for transparency loss
- other models can progressively implement transparency loss

**Comment by f-dy** [comment]
Hum, this is not really implementing transparency carving, is it? Shouldn't there be a loss on the accumulation? What prevents solid stuff of the same color as the background to be reconstructed outside of the masks?

**Comment by nepfaff** [comment]
The background color is random. The 

> Hum, this is not really implementing transparency carving, is it? Shouldn't there be a loss on the accumulation? What prevents solid stuff of the same color as the background to be reconstructed outside of the masks?

That is correct. I missed that before...
There shouldn't be a loss on the accumulation (as is done for a constant background color), but the background color should be added to both the target pixels (as done in this PR) and to the network prediction (not done yet). This then encourages the network to predict zero density to let the background color shine through.
I'm now quite surprised why this works so well with the mustard dataset and the `instant-ngp-bounded` model, even if this part of the implementation is missing. Rendering the accumulation in the viewer, the model indeed learns zero density where alpha is zero.

instantNGP does this [here](https://github.com/NVlabs/instant-ngp/blob/HEAD/src/testbed_nerf.cu#L1499).

**Comment by SamDM** [comment]
Hi, I'd like to clear out a few things:

## About alpha carving:

> There shouldn't be a loss on the accumulation (as is done for a constant background color), but the background color should be added to both the target pixels (as done in this PR) and to the network prediction (not done yet). This then encourages the network to predict zero density to let the background color shine through.

Indeed, in both the ray color prediction as the target image, you need a matching random color that changes in every training evaluation. 

- setting the random background color for the network prediction (i.e. rendering) step happens [here](https://github.com/SamDM/nerfstudio/blob/7776355651a16f1e90242e083166f5ae47156e17/nerfstudio/models/nerfacto.py#L280C1-L285).
- setting the matching background color for the target image happens [here](https://github.com/SamDM/nerfstudio/blob/7776355651a16f1e90242e083166f5ae47156e17/nerfstudio/models/nerfacto.py#L335).

I have also tested penalizing the accumulation on transparent pixels, and got about the same results. The only downside of penalizing accumulation compared to using changing background colors is that you introduce another hyperparam for setting the weight of the accumulation loss.

## My training settings:

The options I used for training were:
```
--pipeline.datamanager.camera-optimizer.mode off
--pipeline.model.disable-scene-contraction True
--pipeline.model.background-color white
--viewer.image-format png
--logging.local-writer.max-log-size 0
nerfstudio-data
--scale-factor 1.0
--scene-scale 0.12
--center-method none
--orientation-method none
--auto-scale-poses False
--train-split-fraction 1.0
```

Also, you must add `use_alpha_channel=True` as a global option to the `cameras.json` file.

Unfortunately, I am not allowed to share the training data, as it is private company data.

**Comment by nepfaff** [comment]
@SamDM You're right. I missed the rendering argument :) Looks good to me functionality-wise!

Code quality-wise, I agree with @f-dy's points. What do you think?

> One concern: the option names and documentation. `use_alpha_channel` is a dataloader option, but we don't know what alpha is being used for. And the transparency loss is implemented only in nerfacto, but undocumented.
> 
> Alternative:
> 
> * alpha channel should be always loaded (so no use_alpha_channel option)
> * models that don't implement transparency loss should discard it
> * nerfacto should have an option to specifically use alpha for transparency loss
> * other models can progressively implement transparency loss

**Comment by nepfaff** [comment]
@SamDM I'd be willing to create a new PR with these changes if you don't want to work on it. Just let me know :) It would be nice to get this merged to be up to date with main.

File: nerfstudio/models/nerfacto.py [hunk 2]
```diff
@@ -288,14 +276,21 @@ def get_outputs(self, ray_bundle: RayBundle):
...
         ray_samples_list.append(ray_samples)
 
-        rgb = self.renderer_rgb(rgb=field_outputs[FieldHeadNames.RGB], weights=weights)
+        if self.kwargs["metadata"].get("use_alpha_channel", False) and self.training:
+            background_color = torch.rand((1, 3))
+        else:
+            background_color = None
+        rgb = self.renderer_rgb(
+            rgb=field_outputs[FieldHeadNames.RGB], weights=weights, background_color=background_color
+        )
```
Comment (jkulhanek) [inline nerfstudio/models/nerfacto.py:285]: I would suggest to use:
```
background_color = self.renderer_rgb.get_background_color(field_outputs[FieldHeadNames.RGB][:, 0, ...].shape)
rgb = self.renderer_rgb(
      rgb=field_outputs[FieldHeadNames.RGB], weights=weights, background_color=background_color
)
```

File: nerfstudio/models/nerfacto.py [hunk 3]
```diff
@@ -327,15 +322,20 @@ def get_outputs(self, ray_bundle: RayBundle):
...
 
     def get_loss_dict(self, outputs, batch, metrics_dict=None):
         loss_dict = {}
         image = batch["image"].to(self.device)
-        loss_dict["rgb_loss"] = self.rgb_loss(image, outputs["rgb"])
+        if outputs.get("background_color", None) is not None and image.shape[-1] == 4:
+            background_color = outputs["background_color"]
+            image_ = image[:, :3] * image[:, 3:] + background_color.to(image.device) * (1 - image[:, 3:])
+        else:
+            image_ = image[:, :3]
```
Comment (jkulhanek) [inline nerfstudio/models/nerfacto.py:337]: Can you please move the code to the renderer and call `self.renderer_rgb.blend_background(...)`?

File: nerfstudio/data/dataparsers/nerfstudio_dataparser.py [hunk 4]
```diff
@@ -296,16 +292,23 @@ def _generate_dataparser_outputs(self, split="train"):
             applied_scale = float(meta["applied_scale"])
             scale_factor *= applied_scale
 
+        if "use_alpha_channel" in meta:
```
Comment (jkulhanek) [inline nerfstudio/data/dataparsers/nerfstudio_dataparser.py:295]: I think this can be removed from the dataparser and the code would always load the alpha-channel if present (only when alpha_color is set it will blend before training - not sure if it will be necessary though).

File: nerfstudio/data/dataparsers/base_dataparser.py [hunk 5]
```diff
@@ -57,6 +57,8 @@ class DataparserOutputs:
     """Camera object storing collection of camera information in dataset."""
     alpha_color: Optional[Float[Tensor, "3"]] = None
     """Color of dataset background."""
+    use_alpha_channel: bool = False
```
Comment (jkulhanek) [inline nerfstudio/data/dataparsers/base_dataparser.py:60]: Please always keep alpha_channel if present and alpha_color is None.

**Comment by SamDM** [comment]
> @SamDM I'd be willing to create a new PR with these changes if you don't want to work on it. Just let me know :) It would be nice to get this merged to be up to date with main.

Hi, I don't have much time to work on this now, so please go ahead, I'd be much obliged!
